ui-src/index.js

```diff
@@ -53,15 +53,41 @@ function req(url) {
   });
 }
 
-// Produces a human-readable but unambiguous format of the given timestamp:
-// ISO-8601 plus an extra colon component after the seconds indicating the
-// fractional time in 90,000ths of a second.
-function formatTime(ts90k) {
-  const m = moment.tz(ts90k / 90, zone);
-  const frac = ts90k % 90000;
-  return m.format('YYYY-MM-DDTHH:mm:ss:' + String(100000 + frac).substr(1) + 'Z');
+/**
+ * Produces a human-readable timestamp in 90k units.
+ *
+ * The format is ISO-8601 plus an extra colon component after the seconds
+ * indicating the fractional time in 90,000ths of a second.
+ *
+ * The timestamp in seconds would be ts90k / 90000, or in ms it would
+ * be ts90k / 90. We use "moment" for formatting, but since it takes only
+ * ms resolution (truncated) we have to be smarter than that to handle
+ * additional precision provided by the timestamp, if the format indicates such
+ * preision is requested.
+ *
+ * We accept a format string in every way compatible with those accepted
+ * by moment's format function, except that we treat the sub-second format
+ * special and extend the functionality to work correctly for precisions up
+ * to 6 decimals (micro-seconds). The default format uses 5 positions.
+ *
+ * @param {int} ts90k timestamp in 90,000ths of a second resolution
+ * @param {string} format Format string
+ * @return {string}
+ */
+function formatTime(ts90k, format = 'YYYY-MM-DDTHH:mm:ss:FFFFFFZ') {
+  const ms = ts90k / 90.0;
+  let msRound = 0.5;
+  const fractionFormat = 'FFFFF';
+  const fmtRegex = new RegExp('(' + fractionFormat + ')', 'g');
+  if (format.match(fmtRegex)) {
+    const frac = ts90k % 90000;
+    format = format.replace(fmtRegex, String(100000 + frac).substr(1));
+    msRound = 0;
+  }
+  return moment.tz(ms + msRound, zone).format(format);
```

Comment:
I don't think it's a mistake that moment.tz truncates rather than rounds to nearest. They probably put some thought into their formatting, and I think when you're looking at a time one of the most important things to know is if it's crossed some threshold or not, and rounding makes that hard. For example, a clock that has a precision of one minute will truncate. I'd think the clock was off if it changed at 30 seconds before the minute rather than right on it.

Reply:
You the rounding was necessary in earlier code where I allowed the S format to go to 6 positions.

A consideration, however, is when you do not use FFFFF, but rather like to use .SSS (producing milliseconds). The timestamp might be having an FFFFF part of 89 which is, effectively almost a millisecond. Without the rounding, the .SSS format would produce ".000", but ".001" would be much closer to the truth.

Point really is that we're formatting for display here with a scenario where we have more precision than moment can handle. We know that moment truncates (why doesn't matter), but without rounding we would therefore produce truncated results.

For now I have produced a non-rounding version. If you agree, we can put rounding back in a future PR.